File: src/lox.ts
import Scanner from './scanner';
import Parser from './parser';
import Interpreter from './interpreter';
import Debug from './debug';
import type Environment from './environment';
import { NativeClock } from './native';

const debug = new Debug('lox').init();

export class Lox {
  public run(text: string, env: Environment) {
    const scanner = new Scanner(text);
    const tokens = scanner.scanTokens();
    debug(tokens);
    const parser = new Parser(tokens);
    const statements = parser.parse();
    debug(statements);
    const interpreter = new Interpreter();
    env.define('clock', new NativeClock())
    interpreter.interpret(statements, env);
  }
}

File: lox.go
package main

import "time"

func interpret(source string, environment *Environment) {
	scanner := NewScanner(source)
	tokens := scanner.ScanTokens()

	parser := NewParser(tokens)
	statements := parser.Parse()

	environment.define("clock", GlobalClock(time.Now().UnixMilli()))

	interpreter := NewInterpreter(environment)
	interpreter.Interpret(statements)

	scanner = nil
	parser = nil
	interpreter = nil
}
